File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/C/aa2d2d5c956e42ab3ac1366954353738.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <time.h>

typedef long long li;
typedef long double ld;
#define AL 26
#define N 1000000 + 13
#define INF (int)(1e9)
#define INF64 (li)(1e18)
#define MOD (INF + 7)
#define EPS 1e-9

int n;
char s[N];
int m;
int t[N];

int nxt[N][AL];
int cnt[N], mx[N];
int masks[3274813];

int read() {
    if (scanf("%s", s) != 1)
        return 0;
    n = strlen(s);
    scanf("%d", &m);
    for (int i = 0; i < m; i++) {
        char buf[N];
        scanf("%s", buf);
        int x = 0;
        for (int j = 0; j < strlen(buf); j++)
            x |= (1 << (buf[j] - 'a'));
        t[i] = x;
    }
    return 1;
}

void solve() {
    for (int i = 0; i < AL; i++) nxt[n][i] = n;
    for (int i = n - 1; i >= 0; --i) {
        for (int j = 0; j < AL; j++)
            nxt[i][j] = nxt[i + 1][j];
        nxt[i][s[i] - 'a'] = i;
    }

    memset(masks, -1, sizeof(masks));
    int mask_count = 0;
    for (int i = 0; i < m; i++) {
        if (masks[t[i]] == -1) {
            masks[t[i]] = mask_count++;
        }
    }

    memset(cnt, 0, sizeof(cnt));
    memset(mx, -1, sizeof(mx));

    for (int i = 0; i < n; i++) {
        int lst = i;
        int cur = (1 << (s[lst] - 'a'));
        while (1) {
            int mn = n;
            for (int j = 0; j < AL; j++) {
                if (!((cur >> j) & 1))
                    mn = fmin(mn, nxt[lst][j]);
            }
            if (masks[cur] != -1) {
                int x = masks[cur];
                if (mx[x] != mn - 1) {
                    mx[x] = mn - 1;
                    ++cnt[x];
                }
            }
            if (mn == n)
                break;
            lst = mn;
            cur |= (1 << (s[lst] - 'a'));
        }
    }

    for (int i = 0; i < m; i++)
        printf("%d\n", cnt[masks[t[i]]]);
}

int main() {
    freopen("input.txt", "r", stdin);

    int tt = clock();
    while (read()) {
        solve();
        fprintf(stderr, "TIME = %d\n", clock() - tt);
        tt = clock();
    }
    return 0;
}

File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/Go/aa2d2d5c956e42ab3ac1366954353738.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

const (
	INF     = int(1e9)
	INF64   = int64(1e18)
	MOD     = INF + 7
	EPS     = 1e-9
	PI      = 3.141592653589793
	N       = 1000*1000 + 13
	AL      = 26
)

var (
	n   int
	s   string
	m   int
	t   [N]int
	nxt [N][AL]int
	cnt [N]int
	mx  [N]int
	masks = make(map[int]int)
)

func read() bool {
	var buf [N]byte
	if _, err := fmt.Scanf("%s", &buf); err != nil {
		return false
	}
	s = string(buf[:])
	n = len(s)
	fmt.Scan(&m)
	for i := 0; i < m; i++ {
		if _, err := fmt.Scanf("%s", &buf); err != nil {
			return false
		}
		tmp := string(buf[:])
		x := 0
		for j := 0; j < len(tmp); j++ {
			x |= (1 << (tmp[j] - 'a'))
		}
		t[i] = x
	}
	return true
}

func solve() {
	for i := 0; i < AL; i++ {
		nxt[n][i] = n
	}
	for i := n - 1; i >= 0; i-- {
		for j := 0; j < AL; j++ {
			nxt[i][j] = nxt[i+1][j]
		}
		nxt[i][s[i]-'a'] = i
	}

	masks = make(map[int]int)
	for i := 0; i < m; i++ {
		if _, exists := masks[t[i]]; !exists {
			k := len(masks)
			masks[t[i]] = k
		}
	}

	for i := range cnt {
		cnt[i] = 0
	}
	for i := range mx {
		mx[i] = -1
	}

	for i := 0; i < n; i++ {
		lst := i
		cur := (1 << (s[lst] - 'a'))
		for {
			mn := n
			for j := 0; j < AL; j++ {
				if (cur>>j)&1 == 0 {
					mn = min(mn, nxt[lst][j])
				}
			}
			if x, exists := masks[cur]; exists {
				if mx[x] != mn-1 {
					mx[x] = mn - 1
					cnt[x]++
				}
			}
			if mn == n {
				break
			}
			lst = mn
			cur |= (1 << (s[lst] - 'a'))
		}
	}

	for i := 0; i < m; i++ {
		fmt.Println(cnt[masks[t[i]]])
	}
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	rand.Seed(time.Now().UnixNano())

	for read() {
		solve()
	}
}
